From: Dimitri Sivanich <sivanich@sgi.com>
Subject: mm: Avoid priority inversion between high priority thread and IO completion
Patch-mainline: no
References: bnc#699916

This change addresses a priority inversion issue between an RT thread
spinning in __lock_page and a page writeback completion thread.

Signed-off-by: Dimitri Sivanich <sivanich@sgi.com>
Signed-off-by: Mel Gorman <mgorman@suse.de>

---
 mm/filemap.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: linux/mm/filemap.c
===================================================================
--- linux.orig/mm/filemap.c	2011-06-14 09:59:22.349637676 -0500
+++ linux/mm/filemap.c	2011-06-14 10:09:17.657726755 -0500
@@ -713,7 +713,8 @@
 	DEFINE_WAIT_BIT(wait, &page->flags, PG_locked);
 
 	do {
-		while(PageUptodate(page) && !need_resched()) {
+		while (PageUptodate(page) && !PageWriteback(page) &&
+				!need_resched()) {
 			cpu_relax();
 			if (!PageLocked(page) && trylock_page(page))
 				goto done;
